### PR TITLE
feat(DTFS2-7543): add e2e deal cancellation maker unable issue facility

### PR DESCRIPTION
**Pull request by rkontogianni**
## Introduction :pencil2:
add e2e deal cancellation maker unable issue facility

### CODE REVIEW

File: e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  });
+
+  after(() => {
+    cy.clearCookies();
+    cy.clearCookie('dtfs-session');
+    cy.clearCookie('_csrf');
+    cy.getCookies().should('be.empty');
+  });
+
+  describe('Deal cancellations on tfm with effective dates in the past and in the future', () => {
```
Comment (MarRobSoftwire) [inline e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js:42]: Could you split this into two describe blocks:

```
describe('when the effective date is in the past', () => {

});

describe('when the effective date is in the future', () => {

});
```

then the `it` descriptions inside each wont need to mention the effective date.

File: e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        cy.clearCookie('dtfs-session');
+        cy.clearCookie('_csrf');
+        cy.getCookies().should('be.empty');
+      });
+    });
+
+    it('AIN deal with unissued facilities is submitted to UKEF, user schedule cancellation in the future in TFM. Maker unable to issue facility on portal', () => {
+      portalPages.contract.visit(deals[0]);
+      portalPages.contract.bondTransactionsTable.row(dealFacilities[1]._id).issueFacilityLink().should('not.exist');
```
Comment (MarRobSoftwire) [inline e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js:66]: Similar to my comment on https://github.com/UK-Export-Finance/dtfs2/pull/3938, I think `deals[0]` is confusing - it would be better to either use `deals.filter(({ tfm }) => tfm.cancellation.effectiveDate >= today )[0]` or, maybe easier to set these as constants at the top rather than using an array at all.

E.g. `const dealWithUnissuedFaciltitiesAndFutureCancellation` (quite a mouthful but more clear? Happy to think about better options too)

File: e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json
Comment (MarRobSoftwire) [inline e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json:1]: We dont normally store mock data as json. Could you save this as a js file?

I think I'd prefer a name like `dealWithoutDates`?

Do we not have any existing test data that could be used for this (with the new dates added to it?)

Comment (ttbarnes) [reply e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json:1]: I wouldn't worry too much about this - we have another (long running draft PR) that replaces things like this and goes through the actual user journey, instead of chucking things into the DB.

https://github.com/UK-Export-Finance/dtfs2/pull/3594

@vamshidharukef There will be conflicts once this is merged

File: e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealReadyToSubmit.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+const dealThatJustNeedsDates = require('./dealThatJustNeedsDates.json');
+const { today } = require('../../../../../../../e2e-fixtures/dateConstants');
+
+module.exports = () => {
```
Comment (MarRobSoftwire) [inline e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealReadyToSubmit.js:4]: Could this be a named export?


```suggestion
const generateDealWithDates = () => {
};

module.exports = {
  generateDealWithDates,
}
```

File: e2e-tests/ukef/cypress/support/commands.js [hunk 4]
```diff
@@ -25,6 +27,9 @@ Cypress.Commands.add('resetPortalUserStatusAndNumberOfSignInLinks', require('./p
 Cypress.Commands.add('enterUsernameAndPassword', require('./portal/enterUsernameAndPassword'));
 Cypress.Commands.add('completeDateFormFields', require('./portal/completeDateFormFields'));
 
+Cypress.Commands.add('makerSubmitDealForReview', makerSubmitDealForReview);
+Cypress.Commands.add('checkerSubmitDealToUkef', checkerSubmitDealToUkef);
```
Comment (MarRobSoftwire) [inline e2e-tests/ukef/cypress/support/commands.js:31]: Nice work - these are really great helpers.

Could the name reflect that they're explicitly `BSS/EWCS` rather than `GEF`

If its not too much of a pain could you add these to the other deal cancellation tests :)

File: e2e-tests/ukef/cypress/support/portal/makerSubmitDealForReview.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+const { BANK1_MAKER1 } = MOCK_USERS;
+
+export const makerSubmitDealForReview = (deal) => {
+  cy.login(BANK1_MAKER1);
+  portalPages.contract.visit(deal);
+  portalPages.contract.proceedToReview().click();
+
+  cy.keyboardInput(portalPages.contractReadyForReview.comments(), 'go');
+  portalPages.contractReadyForReview.readyForCheckersApproval().click();
+};
```
Comment (ttbarnes) [inline e2e-tests/ukef/cypress/support/portal/makerSubmitDealForReview.js:13]: Nice. This and `checkerSubmitDealToUkef` will be great and help in this [draft/WIP PR](https://github.com/UK-Export-Finance/dtfs2/pull/3594/files#diff-caed56733578614ed227f6f8d0f10bb30594077a1fb93741b68f139782c7e9dbR38) @vamshidharukef 

Can we add some documentation please

File: e2e-tests/ukef/cypress/support/portal/checkerSubmitDealToUkef.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import portalPages from '../../../../portal/cypress/e2e/pages';
+import MOCK_USERS from '../../../../e2e-fixtures/portal-users.fixture';
+
+const { BANK1_CHECKER1 } = MOCK_USERS;
+
+export const checkerSubmitDealToUkef = (deal) => {
```
Comment (ttbarnes) [inline e2e-tests/ukef/cypress/support/portal/checkerSubmitDealToUkef.js:6]: Can we add some documentation please

File: e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json
Comment (rkontogianni) [reply e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json:1]: @MarRobSoftwire  we do the same in other files inside the test-data folder, should I change those too ?

File: e2e-tests/ukef/cypress/support/portal/makerSubmitDealForReview.js (hunk 5 above)
Comment (rkontogianni) [reply e2e-tests/ukef/cypress/support/portal/makerSubmitDealForReview.js:13]: @ttbarnes Do you mean comments what kind of documentation do you mean ?

File: e2e-tests/ukef/cypress/support/commands.js (hunk 4 above)
Comment (rkontogianni) [reply e2e-tests/ukef/cypress/support/commands.js:31]: @MarRobSoftwire I'll do that, but can I stick with this name for BSS and makerSubmitGefDealForReview checkerSubmitGefDealToUkef for Gef? Most of the functions are like that

File: e2e-tests/ukef/cypress/support/portal/makerSubmitDealForReview.js (hunk 5 above)
Comment (ttbarnes) [reply e2e-tests/ukef/cypress/support/portal/makerSubmitDealForReview.js:13]: E.g

```js
/**
 * Login as a maker and submit a deal for review
 * @param ...
 */
```

Some fuller exmples here 😊  https://github.com/UK-Export-Finance/dtfs2/tree/main/e2e-tests/tfm/cypress/support/click-events

File: e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js (hunk 2 above)
Comment (rkontogianni) [reply e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js]: I made some changes on this. I created the cy.getOneDeal function in order to have the deal status updated after the cancellation and then in `it`  i use `deals.find((deal) => deal.submissionType === 'Automatic Inclusion Notice' && deal.status === 'Cancelled');` to find the deal and `ainDealPast.facilities.find((facility) => facility.type === 'Bond')._id;` to find the facility.

Comment (MarRobSoftwire) [reply e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js]: This feels a lot more robust - thank you :)

File: e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import generateMinUnissuedDealWithDates from '../../test-data/MIN-deal-unissued-facilities/dealReadyToSubmit';
+import { TFM_URL, PIM_USER_1 } from '../../../../../../../e2e-fixtures';
+
+import { yesterday, tomorrow } from '../../../../../../../e2e-fixtures/dateConstants';
+
+const { BANK1_MAKER1 } = MOCK_USERS;
+
+context('Deal cancellation', () => {
+  const ainUnissuedDeals = Array(2).fill(generateAinUnissuedDealWithDates());
+  const minUnissuedDeals = Array(2).fill(generateMinUnissuedDealWithDates());
```
[the remark this answers is not in the document]
Comment (rkontogianni) [reply e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js:13]: yes it is a draft deal. I gave this name because this deal contains unissued facilities.
`dtfs2/e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json` in this file line 101 mockFacilities.

**Review by MarRobSoftwire — DISMISSED** [review]

File: e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+import MOCK_USERS from '../../../../../../../e2e-fixtures/portal-users.fixture';
+import { generateAinDealUnissuedFacilitiesWithDates } from '../../test-data/AIN-deal-unissued-facilities/dealReadyToSubmit';
+import generateMinDealUnissuedFacilitiesWithDates from '../../test-data/MIN-deal-unissued-facilities/dealReadyToSubmit';
+
+import { TFM_URL, PIM_USER_1 } from '../../../../../../../e2e-fixtures';
+import { yesterday, tomorrow } from '../../../../../../../e2e-fixtures/dateConstants';
+
+const { BANK1_MAKER1 } = MOCK_USERS;
+
+context('Deal cancellation', () => {
```
Comment (ttbarnes) [inline e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js:12]: Can we add a description to this please it's too generic

File: e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    });
+
+    it('MIN deal with unissued facilities is submitted to UKEF, user schedule cancellation deal in the future in TFM. Maker unable to issue facility on portal', () => {
+      const minDealFuture = deals.find((deal) => deal.submissionType === DEAL_SUBMISSION_TYPE.MIN && deal.status === DEAL_STATUS.PENDING_CANCELLATION);
+      const minDealFacilityFuture = minDealFuture.facilities.find((facility) => facility.type === FACILITY_TYPE.BOND)._id;
+      portalPages.contract.visit(minDealFuture);
+      portalPages.contract.bondTransactionsTable.row(minDealFacilityFuture).issueFacilityLink().should('not.exist');
+    });
+  });
+});
```
Comment (ttbarnes) [inline e2e-tests/ukef/cypress/e2e/journeys/portal/deal-cancellation/BSS-EWCS/deal-cancellation-maker-unable-issue-facility.spec.js:87]: Finding this hard to follow, can we split up into something like this please?

```js
describe('when a deal has unissued facilities submitted to UKEF and the deal has already been cancelled', () => {
  describe('AIN deal', () => {
    it('should not allow a Maker to issue facilities in portal', () => { });
  });

  describe('MIN deal', () => {
    it('should not allow a Maker to issue facilities in portal', () => { });
  });
});
```

File: e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealReadyToSubmit.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  deal.mockFacilities[0]['conversionRateDate-day'] = nowDay;
+  deal.mockFacilities[0]['conversionRateDate-month'] = nowMonth;
+  deal.mockFacilities[0]['conversionRateDate-year'] = nowYear;
+
+  deal.mockFacilities[1]['conversionRateDate-day'] = nowDay;
+  deal.mockFacilities[1]['conversionRateDate-month'] = nowMonth;
+  deal.mockFacilities[1]['conversionRateDate-year'] = nowYear;
+
+  return deal;
+};
```
Comment (ttbarnes) [inline e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealReadyToSubmit.js:24]: Do we need to create this? I can see 7 repeated instances of this:

```js
['supplyContractConversionDate-day'] = nowDay;
```

File: e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json [hunk 11]
```diff
@@ -0,0 +1,159 @@
+{
```
Comment (ttbarnes) [inline e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json:1]: Same comment as above, we have 7 instances of `dealThatJustNeedsDates.json`?

File: e2e-tests/ukef/cypress/support/portal-api/getOneDeal.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+const { getDeal, logIn } = require('./api');
+
+/**
+ * User get a deal on portal
```
Comment (ttbarnes) [inline e2e-tests/ukef/cypress/support/portal-api/getOneDeal.js:4]:
```suggestion
 * Get a deal from the portal API
```

As it's not a user getting a deal technically 😄 

Is this for BSS/EWCS and GEF?

**Review by ttbarnes — CHANGES_REQUESTED** [review]
Nice work, few minor comments

File: e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json (hunk 11 above)
Comment (rkontogianni) [reply e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealThatJustNeedsDates.json:1]: @ttbarnes this is different from other ones. it's AIN deal with unissued facilities

File: e2e-tests/ukef/cypress/support/portal-api/getOneDeal.js (hunk 12 above)
Comment (rkontogianni) [reply e2e-tests/ukef/cypress/support/portal-api/getOneDeal.js:4]: @ttbarnes yes it is for both and it blocks me the other pull request I have opened

File: e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealReadyToSubmit.js (hunk 10 above)
Comment (rkontogianni) [reply e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealReadyToSubmit.js:24]: @ttbarnes  There was no mock data for AIN unissued facilities so I created these files. I think if we want to change this approach that needs to be done in a separate ticket. My ticket was only for adding an e2e test and I generally followed the same approach with the rest of the e2e tests.

Comment (ttbarnes) [reply e2e-tests/ukef/cypress/e2e/journeys/portal/test-data/AIN-deal-unissued-facilities/dealReadyToSubmit.js:24]: OK makes sense, thanks

**Review by ttbarnes — APPROVED** [review]
